include/AIToolbox/MDP/TypeTraits.hpp

```diff
@@ -23,16 +23,16 @@ namespace AIToolbox::MDP {
      *
      * @tparam M The class to test for the interface.
      */
-    template <typename M>
+    template <typename M, typename ST, typename AT>
     struct is_generative_model {
         private:
             template <typename Z> static constexpr auto test(int) -> decltype(
 
                     static_cast<size_t (Z::*)() const>                                      (&Z::getS),
                     static_cast<size_t (Z::*)() const>                                      (&Z::getA),
                     static_cast<double (Z::*)() const>                                      (&Z::getDiscount),
-                    static_cast<std::tuple<size_t, double> (Z::*)(size_t,size_t) const>     (&Z::sampleSR),
-                    static_cast<bool (Z::*)(size_t) const>                                  (&Z::isTerminal),
+                    static_cast<std::tuple<ST, double> (Z::*)(ST,AT) const>                 (&Z::sampleSR),
```

Comment:
Two questions in one:
- why not a pair instead of a tuple?
- why not returning in the tuple whether the state is terminal or not as well instead of relying only on ```isTerminal```?
In my case it would be helpful because I need to engage some computation to know whether a state is terminal or not, computation that are done anyway in ```sampleSR```.

Reply:
Also, all the parameters ```ST``` and ```AT``` can be transformed in ```const ST &``` and ```const AT &``` to avoid copying if that's a big object.

Reply:
- I don't use a pair in case I have to add things to the return tuple (which has happened in other classes before). Otherwise all code using the return type would need to be changed. Also with this I can try to have all sampling method return tuples, so they are consistent with each other.
- Most of my work is with MDPs which don't have a terminal state. To avoid having to return false forever, I decided to have a separate `isTerminal` method. One could argue for one or the other, but this is the one I went to.
- I don't really want to add the templated `sampleSR` here. As I said, the `MDP` domain is for non-variable, size_t state and actions. For now I think you can remove the `is_generative_model` requirement from the templated `MCTS`, and we can see how to do that afterwards (maybe by creating a generic `is_generative_model` somewhere else).
- I agree that if you use template state-action parameter, using `const &` is the better choice.

Reply:
Fair enough, I've put a new trait ```is_templated_generative_model``` and conditioned the MCTS to ```is_templated_generative_model_v<M, ST, AT> || is_generative_model<M>``` and removed back all the templating on ```is_generative_model```.